analyzer/src/main/kotlin/managers/Pip.kt

```diff
@@ -622,6 +628,89 @@ class Pip(
 
         return Package.EMPTY.copy(id = id)
     }
+
+    private fun getInstalledPackagesWithLocalMetaData(
+        virtualEnvDir: File,
+        workingDir: File
+    ): List<Package> {
+        val allPackages = listAllInstalledPackages(virtualEnvDir, workingDir)
+
+        // Invoking 'pip show' once for each package separately is too slow, thus obtain the output for all packages
+        // and split it at the separator lines: "---".
+        val output = runInVirtualEnv(
+            virtualEnvDir,
+            workingDir,
+            "pip",
+            "show",
+            "--verbose",
+            *allPackages.map { it.name }.toTypedArray()
+        ).requireSuccess().stdout
+
+        return output.split("---\n").mapNotNull { parsePipShowOutput(it) }
+    }
+
+    /**
+     * Return the [Identifier]s of all installed packages, determined via the command 'pip list'.
+     */
+    private fun listAllInstalledPackages(virtualEnvDir: File, workingDir: File): Set<Identifier> {
+        val json = runInVirtualEnv(virtualEnvDir, workingDir, "pip", "list", "--format", "json")
+            .requireSuccess()
+            .stdout
+
+        val rootNode = jsonMapper.readTree(json) as ArrayNode
+
+        return rootNode.elements().asSequence().mapTo(mutableSetOf()) {
+            Identifier("PyPI", "", it["name"].textValue(), it["version"].textValue())
+        }
+    }
+
+    /**
+     * Parse the output of 'pip show <package-name> --verbose' to a package.
+     */
+    private fun parsePipShowOutput(output: String): Package? {
+        val map = mutableMapOf<String, MutableList<String>>()
+
+        var previousKey: String? = null
+        output.lines().forEach { line ->
+            if (!line.startsWith(" ")) {
+                val index = line.indexOf(":")
+                if (index < 0) return@forEach
+
+                val key = line.substring(0, index)
+                val value = line.substring(index + 1, line.length).trim()
+
+                if (value.isNotEmpty()) {
+                    map.getOrPut(key, { mutableListOf() }) += value
+                }
+
+                previousKey = key
+            } else {
+                if (previousKey == null) return@forEach
+
+                map.getOrPut(previousKey!!, { mutableListOf() }) += line.trim()
```

Comment:
Is the `!!` necessary here?

Reply:
Yes, I however re-wrote it a bit so that it is not necessary anymore.